src/libraries/Microsoft.PowerFx.Core/Public/Values/TableValue.cs

```diff
@@ -109,20 +109,92 @@ public DValue<RecordValue> Index(int index1)
             }
 
             return DValue<RecordValue>.Of(ArgumentOutOfRange(IRContext));
-        }
-
-        // Index() does standard error messaging and then call TryGetIndex().
-        protected virtual bool TryGetIndex(int index1, out DValue<RecordValue> record)
+        }
+
+        /// <summary>
+        /// With mutation support, lookup the record at the given 1-based index, or return an error value if out of range.
+        /// </summary>
+        /// <param name="index1">1-based index.</param>
+        /// <param name="mutationCopy">copies the element, and the table entry pointing to it, when in a mutation context.</param>
+        /// <returns>The record or an errorValue.</returns>
+        public DValue<RecordValue> Index(int index1, bool mutationCopy)
         {
-            var index0 = index1 - 1;
-            if (index0 < 0)
+            if (TryGetIndex(index1, out var record, mutationCopy))
             {
-                record = null;
-                return false;
+                return record;
             }
 
-            record = Rows.ElementAtOrDefault(index0);
-            return record != null;
+            return DValue<RecordValue>.Of(ArgumentOutOfRange(IRContext));
+        }
+
+        // Index() does standard error messaging and then call TryGetIndex().
+        // Can't mutate through this entry point.
+        // It is OK to just override this overload if the table is not mutable.
+        protected virtual bool TryGetIndex(int index1, out DValue<RecordValue> record)
+        {
+            var index0 = index1 - 1;
+            if (index0 < 0)
+            {
+                record = null;
+                return false;
+            }
+
+            record = Rows.ElementAtOrDefault(index0);
+            return record != null;
+        }
+
+        // Index() does standard error messaging and then call TryGetIndex().
+        // This needs to be overriden to support mutation.
+        protected virtual bool TryGetIndex(int index1, out DValue<RecordValue> record, bool mutationCopy)
+        {
+            if (!mutationCopy)
+            {
+                return TryGetIndex(index1, out record);
+            }
+            else
+            {
+                throw new NotImplementedException();
+            }
+        }
+
+        /// <summary>
+        /// Lookup the first record, or return blank if the table is empty.
+        /// </summary>
+        /// <param name="mutationCopy">copies the element, and the table entry pointing to it, when in a mutation context.</param>
+        /// <returns>The record or blank.</returns>
+        public virtual DValue<RecordValue> First(bool mutationCopy = false)
+        {
+            if (mutationCopy)
+            {
+                return DValue<RecordValue>.Of(ImmutableTable(IRContext));

```

Comment:
Suggestion: rename `ImmutableTable` function to `ImmutableTableError`. It took me a while looking at this code to try to understand why First (and Last) would return a table, only to see down this file that the function returned an error.

Reply:
Good call, done.